File: code/Modules/Moon/Moon.cs
namespace Ephemeral
{
    using System.Collections.Generic;
    using System.IO;
    using System.Net;
    using System.Text.RegularExpressions;

    internal class Moon : ModuleMaster, IModule
    {
        public string GetName()
        {
            return "moon";
        }

        public string Run(Dictionary<string, string> settings, string sourceFolder, WebClient webClient)
        {
            var data = GetMoonData(settings, webClient);
            var javascript = File.ReadAllText(sourceFolder + "template.js")
                .Replace("{{PERCENT}}", data.percent)
                .Replace("{{TYPE}}", data.type)
                .Replace("{{URL}}", data.url)
                .Replace("{{LOCATION}}", settings["location"]);

            return javascript;
        }

        private Data GetMoonData(Dictionary<string, string> settings, WebClient webClient)
        {
            var url = "https://www.timeanddate.com/moon/@" + settings["location"];
            var html = webClient.DownloadString(url);

            return new Data
            {
                percent = Regex.Match(html, "<span id=cur-moon-percent>([\\d\\.]+)%").Groups[1].Value,
                type = Regex.Match(html, "(<a href=\"/astronomy/moon/[\\-\\w]+.html\" title=\"[ \\w]+ on the night between [ \\w]+\">[ \\w]+</a>)").Groups[1].Value,
                url = "https://www.timeanddate.com" + Regex.Match(html, "<img id=cur\\-moon src=\"(/scripts/moon\\.php\\?i=[\\d\\.]+&amp;p=[\\d\\.]+&amp;r=[\\d\\.]+)\"").Groups[1].Value.Replace("&amp;", "&")
            };
        }

        private struct Data
        {
            internal string percent;
            internal string type;
            internal string url;
        }
    }
}

File: docs/modules/moon.js
var moonData = {
	Percent:  '12.2',
	Type:     '<a href="/astronomy/moon/waning-crescent.html" title="Waning Crescent on the night between 9 March and 10 March">Waning Crescent</a>',
	Image:    'https://www.timeanddate.com/scripts/moon.php?i=0.122&p=3.345&r=0.268',
	Location: '11609021'
}

function moon() {
	var box = makeBox(1, 1);
	box
		.append('div')
		.attr('class', 'title')
		.html('&#127769;<a href="https://www.timeanddate.com/moon/@' + sunData.Location + '">Moon</a>')
	box = box
		.append('div')
		.attr('class', 'inside');	
	var inside = box
		.append('div');
	inside
		.append('img')
		.attr('src', moonData.Image)
		.attr('width', '50%')
		.attr('height', '50%');		
	inside
		.append('p')
		.html(moonData.Type + ' (' + moonData.Percent + '%)');	
}

modules.push(moon);